packages/go_router/lib/src/route.dart

```diff
@@ -458,4 +558,771 @@ class ShellRoute extends RouteBase {
+/// the builder or pageBuilder methods of StatefulShellRoute to facilitate this.
+/// The container is created using the provided [ShellNavigationContainerBuilder],
+/// where the List of Widgets represent the Navigators for each branch.
+///
+/// Example:
+/// ```
+/// builder: (BuildContext context, GoRouterState state,
+///     StatefulNavigationShell navigationShell) {
+///   return StatefulNavigationShell(
+///     shellRouteState: state,
+///     containerBuilder: (_, __, List<Widget> children) => MyCustomShell(shellState: state, children: children),
+///   );
+/// }
+/// ```
+class StatefulNavigationShell extends StatefulWidget {
+  /// Constructs an [StatefulNavigationShell].
+  StatefulNavigationShell({
+    required this.shellRouteContext,
+    required GoRouter router,
+    required this.containerBuilder,
+  })  : assert(shellRouteContext.route is StatefulShellRoute),
+        _router = router,
+        currentIndex = _indexOfBranchNavigatorKey(
+            shellRouteContext.route as StatefulShellRoute,
+            shellRouteContext.navigatorKey),
+        super(
+            key:
+                (shellRouteContext.route as StatefulShellRoute)._shellStateKey);
+
+  /// The ShellRouteContext responsible for building the Navigator for the
+  /// current [StatefulShellBranch].
+  final ShellRouteContext shellRouteContext;
+
+  /// The builder for a custom container for shell route Navigators.
+  final ShellNavigationContainerBuilder containerBuilder;
+
+  /// The index of the currently active [StatefulShellBranch].
+  ///
+  /// Corresponds to the index in the branches field of [StatefulShellRoute].
+  final int currentIndex;
+
+  final GoRouter _router;
+
+  /// The associated [StatefulShellRoute].
+  StatefulShellRoute get route => shellRouteContext.route as StatefulShellRoute;
+
+  /// Navigate to the last location of the [StatefulShellBranch] at the provided
+  /// index in the associated [StatefulShellBranch].
+  ///
+  /// This method will switch the currently active branch [Navigator] for the
+  /// [StatefulShellRoute]. If the branch has not been visited before, this
+  /// method will navigate to initial location of the branch (see
+  /// [StatefulShellBranch.initialLocation]).
+  void goBranch(int index) {
+    final StatefulShellRoute route =
+        shellRouteContext.route as StatefulShellRoute;
+    final StatefulNavigationShellState? shellState =
+        route._shellStateKey.currentState;
+    if (shellState != null) {
+      shellState.goBranch(index);
+    } else {
+      _router.go(effectiveInitialBranchLocation(index));
+    }
+  }
+
+  /// Gets the effective initial location for the branch at the provided index
+  /// in the associated [StatefulShellRoute].
+  ///
+  /// The effective initial location is either the
+  /// [StackedShellBranch.initialLocation], if specified, or the location of the
+  /// [StackedShellBranch.defaultRoute].
+  String effectiveInitialBranchLocation(int index) {
```

Comment:
I'm not sure this is something users should need to think about too much. I left a comment in the example to explain more.

Reply:
Agree, if we update `goBranch` as I mentioned in the reply.